sys/registry/registry_store.c

```diff
@@ -0,0 +1,126 @@
+#include <string.h>
+#include <errno.h>
+#include "kernel_defines.h"
+#include "clist.h"
+#include "registry/registry.h"
+#include "assert.h"
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+registry_store_t *save_dst;
+clist_node_t load_srcs;
+
+static void _registry_load_cb(char *name, char *val, void *cb_arg)
+{
+    (void)cb_arg;
+    DEBUG("[registry_store] Setting %s to %s\n", name, val);
+    registry_set_value(name, val);
+}
+
+void registry_store_init(void)
+{
+    load_srcs.next = NULL;
+}
+
+void registry_src_register(registry_store_t *src)
+{
+    assert(src != NULL);
+    clist_rpush(&load_srcs, &(src->node));
+}
+
+void registry_dst_register(registry_store_t *dst)
+{
+    assert(dst != NULL);
+    save_dst = dst;
+}
+
+int registry_load(void)
+{
+    clist_node_t *node = load_srcs.next;
+
+    if (!node) {
+        return -ENOENT;
+    }
+
+    do  {
+        registry_store_t *src;
+        src = container_of(node, registry_store_t, node);
+        src->itf->load(src, _registry_load_cb, NULL);
+    } while (node != load_srcs.next);
+    return 0;
+}
+
+static void _registry_dup_check_cb(char *name, char *val, void *cb_arg)
```

Comment:
the logic below AFAIU is as follows:
- Names not identical => `is_dup = 0`
- Names identical and neither `val` or `dup_arg->val` present => `is_dup = 1`
- Names identical and both `val` and `dup_arg->val` present and identical => `is_dup = 1`
- Names identical and only one or other of `val` and `dup_arg->val` present => `is_dup = 0`  

So, you can compare either just the name, or the name and the value, according to your choosing
This seems to give too much flexibility for an internal function; you're the only one who's using it. So you could just keep the check on both `val` and `name`, to save code size. And reduce some of the other checks too as far as possible and just make sure in the calling function that the arguments you're passing aren't causing any errors.

( I just remembered that elsewhere I suggested removing the duplication check. The above comments would apply only if removing the duplication check and silently copying over is considered a bad idea :) )